nni/algorithms/compression/v2/pytorch/base/scheduler.py

```diff
@@ -0,0 +1,89 @@
+    def generate_task(self) -> Tuple[int, Module, List[Dict], Dict[str, Dict[str, Tensor]]]:
+        """
+        Returns
+        -------
+        Tuple[int, Module, List[Dict], Dict[str, Dict[str, Tensor]]]
+            Return new task id, model under pruning, config list used in this task and pre-masks.
+        """
+        raise NotImplementedError()
+
+    def record_task_result(self, task_id: int, pruned_model: Module, masks: Dict[str, Dict[str, Tensor]], score: float,
+                           up_model_masks: Dict[str, Dict[str, Tensor]]):
+        """
+        Used to record the task result.
+
+        Parameters
+        ----------
+        task_id
+            The id of the finished task.
+        pruned_model
+            The pruned model after `pruning_one_step`.
+        masks
+            The masks should be applied on the pruned model.
+        score
+            The score of the pruning performance in this task.
+        up_model_masks
+            The masks applied on the under pruning model.
+        """
+        raise NotImplementedError()
+
+    def pruning_one_step(self, model: Module, config_list: List[Dict], masks: Dict[str, Dict[str, Tensor]]) \
+            -> Tuple[Module, Dict[str, Dict[str, Tensor]], float, Dict[str, Dict[str, Tensor]]]:
+        """
+        Pruning the model with config list.
+
+        Parameters
+        ----------
+        model
+            The model under pruning.
+        config_list
+            The config list usually identify the layers that want to prune.
+        masks
+            The masks should be applied on the under pruning model.
+
+        Returns
+        -------
+        Tuple[Module, Dict[str, Dict[str, Tensor]], float, Dict[str, Dict[str, Tensor]]]
+            Return the pruned model, the masks on pruned model, the score of the pruned model,
+            the masks on under pruning model.
+        """
+        raise NotImplementedError()
+
+    def get_best_result(self) -> Tuple[int, Module, List[Dict], Dict[str, Dict[str, Tensor]], float]:
+        """
+        Returns
+        -------
+        Tuple[int, Module, List[Dict], Dict[str, Dict[str, Tensor]], float]
+            Return the task result that has the best performance,
+            inculde task id, the pruned model, config list used in this task, the masks on the pruned model and score.
+        """
+        raise NotImplementedError()
+
+    def compress(self) -> Tuple[Module, Dict[str, Dict[str, Tensor]]]:
+        """
+        The pruning schedule main loop.
+
+        Returns
+        -------
+        Tuple[Module, Dict[str, Dict[str, Tensor]]]
+            Return the pruned_model and the masks on it in the last iteration.
+        """
+        task_id, model, config_list, pre_masks = self.generate_task()
+        pruned_model, masks = None, None
+
+        while task_id is not None:
+            pruned_model, masks, score, up_model_masks = self.pruning_one_step(model, config_list, pre_masks)
```

Comment:
Do we want to handle exceptions here? If so, what are the possible scenarios? It seems that `pruning_one_step` can fail for many reasons: the pruner throws an error, the ModelSpeedup fails, or even the disk is full... Do we want to step the entire iterative process for all these scenarios, especially with possible distributed applications in the future in mind?

Reply:
I think it would be impossible to take into account all possibilities. But just to give one instance, maybe we should not fully stop when speedup fails/throws an error in one iteration.

Reply:
As we discussed, this place will not handle errors right now.